File: 5. LinkedLists/create_linked_list/create_linked_list.js
// class Node {
//     constructor(val) {
//       this.val = val;
//       this.next = null;
//     }
// }



//Iterative
var createLinkedList = (values) => {
  if(values.length === 0) return null;
  const head = new Node(values[0]);
  let current = head;
  let index = 1;
  while(index < values.length){
    const newNode = new Node(values[index]);
    current.next = newNode;
    current = newNode
    index++;
  }
  return head;
};

//Recursive
var createLinkedList = (values, i = 0) => {
if (i === values.length) return null;
const head = new Node(values[i]);
head.next = createLinkedList(values, i + 1);
return head;
};


//Iterative - 2
const createLinkedList = (values) => {
  if(values.length === 0) return null;
  let prev = new Node(null);
  let head = prev;
  for(let item of values){
    const newNode = new Node(item);
    prev.next = newNode;
    prev = newNode;
  }
  return head.next;
}


// const linkedListValues = (head, arr=[]) => {
//   if(head === null) return [];
//   arr = [head.val, ...linkedListValues(head.next)]
//   return arr;
// }

// let head = createLinkedList(["h", "e", "y"]);

// let arr3 = linkedListValues(head);
// console.log(`Final List: ${arr3}`);



// COMPLEXITY
// Iterative
// n = length of values
// Time: O(n)
// Space: O(n)

// Recursive
// n = length of values
// Time: O(n)
// Space: O(n)

File: 5. LinkedLists/12. create_linked_list/create_linked_list.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None


def print_list(head):
  output = ""
  while(head.next is not None):
    output += str(head.val) + "->"
    head = head.next
  output += str(head.val)
  print(output)

# Iterative
def create_linked_list(values):

    if(len(values) == 0):
        return None
    
    dummyNode = Node(None)
    prev = dummyNode
    for i in range(len(values)):
        newNode = Node(values[i])
        prev.next = newNode
        prev = newNode

    return dummyNode.next

# Recursive
def create_linked_list1(values, index=0):
    if(len(values) == 0):
        return None
    if(index == len(values)):
        return None
    newNode = Node(values[index])
    newNode.next = create_linked_list(values, index+1)
    return newNode


    
print_list(create_linked_list(["h", "e", "y"]))
# print_list(create_linked_list1(["h", "e", "y"]))